cpp/src/community/louvain_impl.cuh

```diff
@@ -82,11 +84,108 @@ std::pair<std::unique_ptr<Dendrogram<vertex_t>>, weight_t> louvain(
+        std::vector<size_t> tx_value_counts(comm_size, 0);
+
+        for (int i = 0; i < comm_size; i++) {
+          tx_value_counts[i] = random_cluster_assignments.size() / comm_size + i <
+                                   (random_cluster_assignments.size() % comm_size)
+                                 ? 1
+                                 : 0;
+        }
+
+        std::tie(random_cluster_assignments, std::ignore) =
+          cugraph::shuffle_values(handle.get_comms(),
+                                  random_cluster_assignments.begin(),
+                                  tx_value_counts,
+                                  handle.get_stream());
+
+        // shuffle/permute locally
+        random_numbers.resize(random_cluster_assignments.size(), handle.get_stream());
+
+        cugraph::detail::uniform_random_fill(handle.get_stream(),
+                                             random_numbers.data(),
+                                             random_numbers.size(),
+                                             float{0.0},
+                                             float{1.0},
+                                             *rng_state);
+        thrust::sort_by_key(handle.get_thrust_policy(),
+                            random_numbers.begin(),
+                            random_numbers.end(),
+                            random_cluster_assignments.begin());
+
+        // find out how many elements current GPU needs to send to other GPUs
+
+        std::fill(tx_value_counts.begin(), tx_value_counts.end(), 0);
+        auto sample_buffer_sizes = cugraph::host_scalar_allgather(
+          handle.get_comms(), random_cluster_assignments.size(), handle.get_stream());
+
+        auto expected_sample_buffer_sizes = cugraph::host_scalar_allgather(
+          handle.get_comms(), graph_view.local_vertex_partition_range_size(), handle.get_stream());
+
+        std::vector<size_t> nr_smaples_per_GPU(comm_size, 0);
+        for (int i = 0; i < comm_size; i++) {
+          size_t nr_samples_ith_gpu = sample_buffer_sizes[i];
+          for (int j = 0; nr_samples_ith_gpu > 0 && j < comm_size; j++) {
+            if (expected_sample_buffer_sizes[j] > static_cast<vertex_t>(nr_smaples_per_GPU[j])) {
+              size_t delta = std::min(nr_samples_ith_gpu,
+                                      expected_sample_buffer_sizes[j] - nr_smaples_per_GPU[j]);
+              if (comm_rank == i) { tx_value_counts[j] = delta; }
+              nr_smaples_per_GPU[j] += delta;
+              nr_samples_ith_gpu -= delta;
+            }
+          }
+        }
+        // shuffle as many vertices as local vertex partition size to each GPU.
+        std::tie(random_cluster_assignments, std::ignore) =
+          cugraph::shuffle_values(handle.get_comms(),
+                                  random_cluster_assignments.begin(),
+                                  tx_value_counts,
+                                  handle.get_stream());
```

Comment:
You don't need to re-process the entire values.

Say, you count the deficits and extras.

You can run `device_allgatherv` on extras.
You can run exclusive_scan on deficits.
Then, in a GPU with a non-zero deficit, you can determine the subrange of the device_allgatherv output to copy.

This will involve much less traffic. If you want to make this algorithm to scale to many thousands of GPUs, you can replace device_allgatherv with P2P. You can compute which GPUs contain the data for the subrange. Then, you can set-up P2P based on this calculation. This will be necessary if the device_allgatherv output array is comparable or larger than # vertices / # GPUs.